Caching `target` for macos CI build is flaky; Unknown Cargo issue.

https://github.com/apollographql/rust/runs/1038067162?check_suite_focus=true

If we disable caching, everything works.
The action is defined as [on cache miss: compile everything and cache] + [load from cache and test]

The reason for this separation is that we could cache compilation if tests **fail** so that subsequent runs are fast.

When loading `target` from cache, compilation fails with this error:
```
   Compiling graphql-parser v0.3.0 (/Users/runner/work/rust/rust/graphql-parser)
error[E0463]: can't find crate for `thiserror_impl` which `thiserror` depends on
 --> graphql-parser/src/error.rs:2:5
  |
2 | use thiserror::Error;
  |     ^^^^^^^^^ can't find crate

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
```

We currently do not know why this is happening. 
I had suspected some kind of corruption, but Cargo is supposed to be resilient to that. I don't know what conditions would cause Cargo to see this error.

In the interim, the CI actions will be separated to [linux only] and [all] so that we could get quick and reliable feedback from the linux only action. We disabled caching on macos for the slow action.

--- a/.github/workflows/slow-test.yml
+++ b/.github/workflows/slow-test.yml
@@ -42,7 +42,7 @@ jobs:
 
       ### BUILD CACHE ###
       # NB: We install gnu-tar because BSD tar is buggy on Github's macos machines. https://github.com/actions/cache/issues/403
-      - name: Install GNU tar
+      - name: Install GNU tar (Macos)
         if: matrix.build == 'macos'
         run: |
           brew install gnu-tar

--- a/.github/workflows/slow-test.yml
+++ b/.github/workflows/slow-test.yml
@@ -41,10 +41,13 @@ jobs:
           override: true
 
       ### BUILD CACHE ###
+      # NB: We install gnu-tar because BSD tar is buggy on Github's macos machines. https://github.com/actions/cache/issues/403
+      - name: Install GNU tar
+        if: matrix.build == 'macos'
+        run: |
+          brew install gnu-tar
+          echo "::add-path::/usr/local/opt/gnu-tar/libexec/gnubin"
       - name: Cache Cargo registry, target, index
-        # we skip caching on macos because of the flaky `thiserror_impl` problem.
-        # see https://github.com/apollographql/rust/issues/123
-        if: matrix.build != 'macos'
         uses: actions/cache@v2
         with:
           path: |

